refactor(konnect): add SDK interface types and theirs mocks

pmalek opened the pull request:
**What this PR does / why we need it**:

This PR will allow UTs for entity focused functions when mocks for those are introduced.

This PR does reorganize the `konnect` package a bit to solve the import cycles between `konnect` and newly added `ops` package.

Mocks are generated using [`mockery`](https://github.com/vektra/mockery) and `generate.mocks` makefile target. The config that configures which interfaces to generate mocks for is `.mockery.yaml`.

The tests that are added in this PR are in 
- `controller/konnect/ops/ops_controlplane_test.go`
- `controller/konnect/ops/ops_kongservice_test.go`

More will follow after this PR gets reviewed and merged.

**Which issue this PR fixes**

Part of #447

**Special notes for your reviewer**:

pmalek committed:
refactor(konnect): add SDK interface types for each entity type

pmalek committed:
tests: add mockery for Konnect SDK

pmalek committed:
tests: add mocks for KongService SDK

programmer04 commented on .golangci.yaml:
```diff
@@ -62,6 +62,13 @@ linters-settings:
+      - pkg: github.com/Kong/sdk-konnect-go/models/operations
+        alias: sdkkonnectgoops
+      - pkg: github.com/Kong/sdk-konnect-go/models/sdkerrors
+        alias: sdkkonnectgoerrs
```
I think this `go` in the middle is redundant. WDYT e.g. `sdkkonnectops` instead of `sdkkonnectgoops`? The same for other packages

randmonkey replied:
Changed the aliases.

programmer04 reviewed (commented):
Overall LGTM

randmonkey committed:
Resolve conflict for "refactor(konnect): add SDK interface types and theirs mocks"(#507) (#542)

* feat(cli): configurable certgen images (#516)

* feat(cli): configurable certgen images

Add two new command line options to the manager.

  -webhook-certificate-config-base-image string
    	The base image for the certgen Jobs. (default "registry.k8s.io/ingress-nginx/kube-webhook-certgen:v1.3.0")
  -webhook-certificate-config-shell-image string
    	The shell image for the certgen Jobs. (default "busybox")

Those are optional. If you omit them the defaults will be used which are
the previous hard coded values.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* feat(cli): cleanup code per review

Removed unnecessary variable assignments.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* feat(cli): add CHANGELOG entry

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* feat(cli): unit tests handle new arguments

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* feat(cli): move new configuration into Config

Also created a constant for the shell image, so it is no more a hard
coded string hidden in the code.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* feat(cli): added UT for the new arguments

The new unit test tests only if the command line arguments are set in
the configuration. The other cases, the default configuration, the
environmental variable handling are already tested in previous cases.

The test does not cover if the values from the configuration are
actually used but at the moment there are no tests for that part of the
code at all.

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

---------

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>

* chore(deps): bump github.com/kong/kubernetes-configuration (#523)

Bumps [github.com/kong/kubernetes-configuration](https://github.com/kong/kubernetes-configuration) from 0.0.8 to 0.0.9.
- [Commits](https://github.com/kong/kubernetes-configuration/compare/v0.0.8...v0.0.9)

---
updated-dependencies:
- dependency-name: github.com/kong/kubernetes-configuration
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* chore(deps): bump github.com/kong/kubernetes-ingress-controller/v3 from 3.2.4 to 3.3.0 (#531)

* chore(deps): bump github.com/kong/kubernetes-ingress-controller/v3

Bumps [github.com/kong/kubernetes-ingress-controller/v3](https://github.com/kong/kubernetes-ingress-controller) from 3.2.4 to 3.3.0.
- [Release notes](https://github.com/kong/kubernetes-ingress-controller/releases)
- [Changelog](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md)
- [Commits](https://github.com/kong/kubernetes-ingress-controller/compare/v3.2.4...v3.3.0)

---
updated-dependencies:
- dependency-name: github.com/kong/kubernetes-ingress-controller/v3
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* chore: go mod tidy

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jakub Warczarek <jakub.warczarek@konghq.com>

* `KongPluginBinding` reconciler (#513)

* feat: KongPluginBinding reconciler

Signed-off-by: Mattia Lavacca <lavacca.mattia@gmail.com>

* chore: fix linter

Signed-off-by: Mattia Lavacca <lavacca.mattia@gmail.com>

* chore: CHANGELOG updated

Signed-off-by: Mattia Lavacca <lavacca.mattia@gmail.com>

* Update controller/konnect/ops_kongpluginbinding.go

Co-authored-by: Grzegorz Burzyński <czeslavo@gmail.com>

---------

Signed-off-by: Mattia Lavacca <lavacca.mattia@gmail.com>
Co-authored-by: Grzegorz Burzyński <czeslavo@gmail.com>

* chore(deps): update kong/kubernetes-ingress-controller docker tag to v3.3.0 (#532)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* chore(deps): update kong/kubernetes-ingress-controller docker tag to v3.3.1 (#534)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* Add watch for KongPlugin and KongClusterPlugin in KongPluginBinding controller (#535)

* chore(deps): bump github.com/kong/kubernetes-ingress-controller/v3 (#536)

Bumps [github.com/kong/kubernetes-ingress-controller/v3](https://github.com/kong/kubernetes-ingress-controller) from 3.3.0 to 3.3.1.
- [Release notes](https://github.com/kong/kubernetes-ingress-controller/releases)
- [Changelog](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md)
- [Commits](https://github.com/kong/kubernetes-ingress-controller/compare/v3.3.0...v3.3.1)

---
updated-dependencies:
- dependency-name: github.com/kong/kubernetes-ingress-controller/v3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* chore(deps): bump github.com/Masterminds/sprig/v3 in /hack/generators (#539)

Bumps [github.com/Masterminds/sprig/v3](https://github.com/Masterminds/sprig) from 3.2.3 to 3.3.0.
- [Release notes](https://github.com/Masterminds/sprig/releases)
- [Changelog](https://github.com/Masterminds/sprig/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Masterminds/sprig/compare/v3.2.3...v3.3.0)

---
updated-dependencies:
- dependency-name: github.com/Masterminds/sprig/v3
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* chore(deps): update dependency kubernetes-sigs/controller-tools to v0.16.2 (#538)

* chore(deps): update dependency kubernetes-sigs/controller-tools to v0.16.2

* chore: regenerate

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jakub Warczarek <jakub.warczarek@konghq.com>

* chore(deps): bump github.com/gruntwork-io/terratest (#541)

Bumps [github.com/gruntwork-io/terratest](https://github.com/gruntwork-io/terratest) from 0.47.0 to 0.47.1.
- [Release notes](https://github.com/gruntwork-io/terratest/releases)
- [Commits](https://github.com/gruntwork-io/terratest/compare/v0.47.0...v0.47.1)

---
updated-dependencies:
- dependency-name: github.com/gruntwork-io/terratest
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* update golangci configuration

---------

Signed-off-by: Arpad Kunszt <akunszt@hiya.com>
Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Mattia Lavacca <lavacca.mattia@gmail.com>
Co-authored-by: akunszt <32456696+akunszt@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jakub Warczarek <jakub.warczarek@konghq.com>
Co-authored-by: Mattia Lavacca <lavacca.mattia@gmail.com>
Co-authored-by: Grzegorz Burzyński <czeslavo@gmail.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

randmonkey committed:
Merge remote-tracking branch 'origin/main' into konnect-use-sdk-interface

randmonkey committed:
fix lint

randmonkey committed:
Merge remote-tracking branch 'origin/main' into konnect-use-sdk-interface

randmonkey committed:
replace KonnnectControlPlane to KonnectGatewayControlPlane

czeslavo commented on .mockery.yaml:
```diff
@@ -0,0 +1,15 @@
+packages:
+  github.com/kong/gateway-operator/controller/konnect/ops:
+    interfaces:
+      ControlPlaneSDK:
+      ServicesSDK:
```
Nit: we might as well already generate mocks for other interfaces that are added in this PR (`ConsumerGroupSDK`, `ConsumersSDK`, `RoutesSDK`).

randmonkey replied:
I'll add mocks of all used SDKs in the following PR to refactor the usage of SDK.
